File: cpp/sample4.cpp
#include <iostream>
#include <cstdlib>
#include <ctime>

#include <vector>

const int SIZE=64;

int main(int argc, char**argv){
  int i;
  const int ishow=16;
  double sum;
  std::vector<double> datah(SIZE);
  
  printf("CPU:\n");

  for(i=0;i<SIZE;i++) datah[i] = i;

  printf("data: "); for(i=0;i<ishow;i++) printf(" %.2f",datah[i]); printf("...\n");
  
  sum=0.0;
  for(i=0;i<SIZE;i++){
    sum = sum + datah[i];
  }
  
  printf("Numerical Result: "); printf(" %.2f\n",sum);
  
  sum = SIZE*(SIZE-1)/2;
  printf("Analytic  Result: "); printf(" %.2f\n",sum);
  
  
  return 0;
}

File: fortran_openacc/sample4.f90
program main
  implicit none
  real(4)::tbgn,tend
  integer:: i
  integer, parameter:: ishow=16
  integer, parameter:: SIZE=64
  
  real(8),dimension(SIZE):: datah
  real(8):: sum
  
  write(6,*) "CPU:\n"

  
  do i=1, SIZE
     datah(i) = i
  enddo

  write(6,*,advance='no')"data: "
  do i=1,ishow
     write(6,"(1x,f8.2)",advance='no') datah(i)
  enddo
  write(6,*) ""

  sum=0.0d0
!$acc data copyin(sum,datah),copyout(sum)
!$acc parallel loop reduction(+:sum)
  do i=1, SIZE
     sum = sum + datah(i)
  enddo
!$acc end parallel
!$acc end data 
 
  write(6,*,advance='no')"Numerical result: "
  write(6,"(1x,f8.2)") sum
  sum = SIZE*(SIZE+1)/2
  write(6,*,advance='no')"Analytic  result: "
  write(6,"(1x,f8.2)") sum
 
end program
